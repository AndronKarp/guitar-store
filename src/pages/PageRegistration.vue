<template>
  <b-form>
    <custom-input
      v-for="(field, index) in Object.keys(form)"
      :key="index"
      v-model="$v.form[field].value.$model"
      :meta="form[field].meta"
      :validations="$v.form[field].value"
    ></custom-input>
  </b-form>
</template>

<script>
import { validationMixin } from "vuelidate";
import * as validators from "vuelidate/lib/validators";
import CustomInput from "../components/CustomInput";

export default {
  data() {
    return {
      form: {
        name: {
          meta: {
            type: "text",
            placeholder: "Your name...",
            validations: {
              required: {
                rule: validators.required,
                errorMessage: "Required field"
              },
              alpha: {
                rule: validators.alpha,
                errorMessage: "Name must contain only letters"
              },
              minLength: {
                rule: validators.minLength(2),
                errorMessage: "Name must have at least 2 characters"
              }
            }
          },
          value: ""
        },
        email: {
          meta: {
            type: "email",
            placeholder: "Your e-mail...",
            validations: {
              required: {
                rule: validators.required,
                errorMessage: "Required field"
              },
              email: {
                rule: validators.email,
                errorMessage: "Invalid e-mail"
              }
            }
          },
          value: ""
        },
        password: {
          meta: {
            type: "password",
            placeholder: "Your password...",
            validations: {
              required: {
                rule: validators.required,
                errorMessage: "Required field"
              },
              alphaNum: {
                rule: validators.alphaNum,
                errorMessage:
                  "Password must contain only letters and/or numbers"
              },
              minLength: {
                rule: validators.minLength(6),
                errorMessage: "Password must contain at least 6 characters"
              }
            }
          },
          value: ""
        },
        confirmPassword: {
          meta: {
            type: "password",
            placeholder: "Confirm password...",
            validations: {
              required: {
                rule: validators.required,
                errorMessage: "Required field"
              },
              sameAsPassword: {
                rule: validators.sameAs(function() {
                  return this.form.password.value;
                }),
                errorMessage: "Passwords don't match"
              }
            }
          },
          value: ""
        }
      }
    };
  },
  validations() {
    const form = {};
    const formFields = Object.entries(this.form);
    formFields.forEach(([fieldName, field]) => {
      const value = {};
      const validations = Object.entries(field.meta.validations);
      validations.forEach(([validationName, validation]) => {
        value[validationName] = validation.rule;
      });
      form[fieldName] = { value };
    });
    return { form };
  },
  components: {
    CustomInput
  },
  mixins: [validationMixin]
};
</script>

<style></style>
