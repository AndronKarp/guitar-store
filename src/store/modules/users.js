import { auth, usersRef } from "@/configs/firebase";

export default {
  state: {
    currentUser: null,
    users: []
  },
  getters: {
    currentUser(state) {
      return state.currentUser;
    },
    isEmailTaken(state) {
      return email => !!state.users.find(user => user.email === email);
    }
  },
  mutations: {
    setCurrentUser(state, user) {
      state.currentUser = user;
    },
    pushToUsers(state, user) {
      state.users.push(user);
    }
  },
  actions: {
    setAuthObserver(store) {
      auth.onAuthStateChanged(user => {
        store.commit("setCurrentUser", user);
      });
    },
    async updateUserDisplayName({ state, commit }, displayName) {
      await state.currentUser.updateProfile({ displayName });
      commit("setCurrentUser", { ...state.currentUser, displayName });
    },
    fetchUsers(store) {
      usersRef.on("child_added", snapshot => {
        store.commit("pushToUsers", snapshot.val());
      });
    }
  }
};
