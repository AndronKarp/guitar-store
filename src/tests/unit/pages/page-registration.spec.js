import { createLocalVue, mount } from "@vue/test-utils";
import PageRegistration from "@/pages/PageRegistration";
import { BootstrapVue } from "bootstrap-vue";
import Vuelidate from "vuelidate";
import Vuex from "vuex";
import users from "@/store/modules/users";

const localVue = createLocalVue();

localVue.use(BootstrapVue);
localVue.use(Vuelidate);
localVue.use(Vuex);

describe("PageRegistration.vue", () => {
  let wrapper;
  let state;
  let actions;
  let store;

  beforeEach(() => {
    state = { currentUser: null, users: [{ email: "e-mail" }] };
    actions = { updateUserDisplayName: jest.fn() };
    store = new Vuex.Store({
      state,
      getters: users.getters,
      actions
    });
    wrapper = mount(PageRegistration, { localVue, store });
  });

  test("renders registration form", () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe("computed", () => {
    const exampleField = "name";
    let exampleInput;

    describe("validationState", () => {
      beforeEach(() => {
        exampleInput = wrapper.find(`#${exampleField}`);
      });

      test("returns null if input is clear", () => {
        expect(wrapper.vm.validationState(exampleField)).toBe(null);
      });

      test("returns false if input is dirty", () => {
        exampleInput.setValue("a");
        exampleInput.setValue("");
        expect(wrapper.vm.validationState(exampleField)).toBe(false);
      });

      test("returns false if input data is invalid", () => {
        exampleInput.setValue("a");
        expect(wrapper.vm.validationState(exampleField)).toBe(false);
      });

      test("returns true if input data is valid", () => {
        exampleInput.setValue("Andrew");
        expect(wrapper.vm.validationState(exampleField)).toBe(true);
      });
    });

    describe("validationErrorMessage", () => {
      beforeEach(() => {
        exampleInput = wrapper.find(`#${exampleField}`);
      });

      test("returns empty array if input is clear", () => {
        expect(wrapper.vm.validationErrorMessage(exampleField)).toEqual([]);
      });

      test("returns null if input data is valid", () => {
        exampleInput.setValue("Andrew");
        expect(wrapper.vm.validationErrorMessage(exampleField)).toBe(null);
      });

      test("returns error message of failed validation if input data is invalid", () => {
        exampleInput.setValue("a1");
        expect(wrapper.vm.validationErrorMessage(exampleField)).toBe(
          wrapper.vm.form[exampleField].validations.alpha.errorMessage
        );
      });
    });

    describe("form submit button", () => {
      let submitButton;
      let inputs;

      beforeEach(() => {
        submitButton = wrapper.find("button");
        inputs = wrapper.findAll("input");
      });

      test("disabled if form is empty", () => {
        expect(submitButton.element.getAttribute("disabled")).toBe("disabled");
      });

      test("disabled if input data is invalid", () => {
        inputs.setValue("invalid");
        wrapper.vm.$nextTick(() => {
          expect(submitButton.element.getAttribute("disabled")).toBe(
            "disabled"
          );
        });
      });

      test("enabled if input data is valid", () => {
        inputs.wrappers[0].setValue("Andrew");
        inputs.wrappers[1].setValue("andrewkarp00@gmail.com");
        inputs.wrappers[2].setValue("123456");
        inputs.wrappers[3].setValue("123456");
        wrapper.vm.$nextTick(() => {
          expect(submitButton.element.getAttribute("disabled")).toBe(null);
        });
      });
    });
  });
});
