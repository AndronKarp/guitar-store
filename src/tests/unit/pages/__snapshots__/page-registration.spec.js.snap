// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageRegistration.vue renders registration form 1`] = `
<div class="d-flex flex-fill justify-content-center align-items-center">
  <form class="d-flex flex-column w-25" style="min-width: 280px;">
    <fieldset class="form-group" id="__BVID__21">
      <!---->
      <div tabindex="-1" role="group" class="bv-no-focus-ring"><input id="name" type="text" placeholder="Your name..." class="form-control">
        <div class="invalid-feedback">
          []
        </div>
        <!---->
        <!---->
        <!---->
      </div>
    </fieldset>
    <fieldset class="form-group" id="__BVID__23">
      <!---->
      <div tabindex="-1" role="group" class="bv-no-focus-ring"><input id="email" type="email" placeholder="Your e-mail..." class="form-control">
        <div class="invalid-feedback">
          []
        </div>
        <!---->
        <!---->
        <!---->
      </div>
    </fieldset>
    <fieldset class="form-group" id="__BVID__25">
      <!---->
      <div tabindex="-1" role="group" class="bv-no-focus-ring"><input id="password" type="password" placeholder="Your password..." class="form-control">
        <div class="invalid-feedback">
          []
        </div>
        <!---->
        <!---->
        <!---->
      </div>
    </fieldset>
    <fieldset class="form-group" id="__BVID__27">
      <!---->
      <div tabindex="-1" role="group" class="bv-no-focus-ring"><input id="confirm-password" type="password" placeholder="Confirm password..." class="form-control">
        <div class="invalid-feedback">
          []
        </div>
        <!---->
        <!---->
        <!---->
      </div>
    </fieldset> <button type="submit" disabled="disabled" class="btn btn-secondary disabled">Sign Up</button>
  </form>
</div>
`;
